src/bindings/python/flux/constraint/parser.py

```diff
@@ -0,0 +1,453 @@
+        if self.pending_token is not None:
+            t.value = self.pending_token.value + t.value
+            self.pending_token = None
+        elif t.value.endswith(":"):
+            #  Save a token that ends with ':' to possibly combine with
+            #  any following token. This allows op:"quoted string"
+            self.pending_token = t
+            return None
+        return t
+
+    def t_eof(self, t):
+        if self.pending_token is not None:
+            val = self.pending_token.value
+            raise ConstraintSyntaxError(f"Missing argument to token '{val}'")
+        return None
+
+    def t_QUOTE(self, t):
+        r"'|\""  # fmt: skip
+        self.quote_start = t.lexer.lexpos
+        self.last_quote = t.lexer.lexpos - 1
+        if t.value == "'":
+            t.lexer.begin("squoting")
+        else:
+            t.lexer.begin("dquoting")
+
+    #  quoting state:
+    def t_squoting_TOKEN(self, t):
+        r"([^'])+"
+        return self.t_TOKEN(t)
+
+    def t_squoting_QUOTE(self, t):
+        r"'"
+        self.last_quote = None
+        t.lexer.begin("INITIAL")
+
+    def t_squoting_eof(self, t):
+        pos = self.quote_start
+        raise ConstraintSyntaxError(f'Unclosed quote "\'" at position {pos}')
+
+    def t_dquoting_TOKEN(self, t):
+        r"([^\"])+"
+        return self.t_TOKEN(t)
+
+    def t_dquoting_QUOTE(self, t):
+        r"\""  # fmt: skip
+        self.last_quote = None
+        t.lexer.begin("INITIAL")
+
+    def t_dquoting_eof(self, t):
+        pos = self.quote_start
+        raise ConstraintSyntaxError(f"Unclosed quote '\"' at position {pos}")
+
+
+class ConstraintParser:
+    r"""
+    Base constraint query syntax parser class.
+
+    This class implements an RFC 35 compliant constraint query syntax
+    parser with the following simplified grammar:
+
+    expr : expr expr
+     | expr and expr
+     | expr or expr
+     | not expr
+     | '(' expr ')'
+     | '-' term
+     | term
+
+    and : &{1,2}|and|AND
+
+    or : \|{1,2}|or|OR
+
+    not : not|NOT
+
+    term : \w*:?.+  # i.e. [operator:]operand
+
+    Where a term is a constraint operation which has the form
+    '[operator:]operand'. If the token does not include a `:`, then a
+    class default operator may optionally be substituted, e.g. "operand"
+    becomes "default:operand".
+
+    By default, ``operand`` is included as a single entry in the RFC 31
+    values array for the operator, i.e. ``{"operator":["operand"]}``.
+    However, if ``operator`` appears in the self.split_values dict
+    of the parser object, then the corresponding string will be used
+    to split ``value`` into multiple entries. E.g. if
+    ::
+
+        split_values = { "op": "," }
+
+    Then ``op:foo,bar`` would result in ``{"op":["foo","bar"]}``.
+
+    Terms are joined by AND unless OR is specified, such that ``a b c``
+    is the same as ``a && b && c``. A term can be negated with ``-``,
+    but to negate a whole expression, NOT must be used.
```

Comment:
given you have an example for the first statement, perhaps an example(s) for the second statement?